src/Bicep.Core/Semantics/ParamsSemanticModel.cs

```diff
@@ -27,11 +30,96 @@ public ParamsSemanticModel(BicepParamFile bicepParamFile, Compilation? bicepComp
             // name binding is done
             // allow type queries now
             paramsSymbolContext.Unlock();
+            
         }
 
-        public IEnumerable<IDiagnostic> GetDiagnostics()
-            => BicepParamFile.ProgramSyntax.GetParseDiagnostics();
+        /// <summary>
+        /// Gets all the params semantic diagnostics unsorted. Does not include params parser and params lexer diagnostics.
+        /// </summary>
+        /// <returns></returns>
+        private IEnumerable<IDiagnostic> GetAdditionalSemanticDiagnostics()
+        {
+            if (this.BicepCompilation is null)
+            {
+                return Enumerable.Empty<IDiagnostic>();
+            }
+            var requiredParameters = this.BicepCompilation.GetEntrypointSemanticModel().Parameters;
+            var parameterAssignmentSyntax = BicepParamFile.ProgramSyntax.Children.OfType<ParameterAssignmentSyntax>();
+
+            var sortedRequiredParameters =  requiredParameters.OrderBy(x => x.Name);
+            var sortedParameterAssignments = parameterAssignmentSyntax.OrderBy(x => this.ParamBinder.GetSymbolInfo(x)?.Name);
+
+            // get errors relating to missing declarations or assignments
+            AddMissingErrors(sortedRequiredParameters.ToArray(), sortedParameterAssignments.ToArray());
+            // get errors relating to type mismatch of params between Bicep and params files
+            AddTypeMismatchErrors(sortedParameterAssignments);
+
+            return semanticDiagnostics;
+        }
+
+        private void AddMissingErrors(Metadata.ParameterMetadata[] sortedRequiredParameters, ParameterAssignmentSyntax[] sortedParameterAssignments)
+        {
+            var i = 0;
+            var j = 0;
+
+            while ((i < sortedRequiredParameters.Length) && (j < sortedParameterAssignments.Length))
```

Comment:
You could probably simplify this code using `Dictionary` rather than sorted arrays. Note that C# will allow you to provide a comparer for dictionary keys - e.g. `var requiredParamsDict = requiredParameters.ToDictionary(x => x.Name, LanguageConstants.IdentifierComparer);`

This would allow you to e.g. find missing required params using Linq:
```csharp
var missingRequiredParams = requiredParamsDict.Values.Where(x => !assignedParamsDict.ContainsKey(x.Name));
```

Reply:
Great refactor, this is looking really clean and easy to follow 👍